### DISCUSSION TITLE
What factors affect search in queue latency?

### DISCUSSION THREAD

**Discussion by redlion0929** [milvus-io/milvus#22075] [Q&A and General discussion] [answered]
Hi. 
I am deploying and using Milvus Cluster on EKS, and my current specifications are as follows.
```
* rootCoordinator : 1 (limits cpu:1, memory : 2Gi)
* proxy : 1 (limits cpu:1, memory:4Gi) 
* index
    * indexCoordinator : 1 (limits cpu:0.5, memory: 0.5Gi)
    * indexNode: 2 (limits cpu:4, memory: 8Gi)
* query
    * queryCoordinator : 1 (limits cpu:0.5, memory: 0.5Gi)
    * queryNode: 5 (limits cpu:4, memory: 16Gi)
* data
    * dataCoordinator : 1 (limits cpu:0.5, memory: 0.5Gi)
    * dataNode: 1 (limits cpu:1, memory: 4Gi)
* etcd:
  autoCompactionMode: revision
  autoCompactionRetention: "1000"
  extraEnvVars:
  - name: ETCD_QUOTA_BACKEND_BYTES
    value: "4294967296"
  - name: ETCD_HEARTBEAT_INTERVAL
    value: "500"
  - name: ETCD_ELECTION_TIMEOUT
    value: "25000"
  - name: ETCD_SNAPSHOT_COUNT
    value: "10000"
  - name: ETCD_ENABLE_PPROF
    value: "true"
  persistence:
    accessMode: ReadWriteOnce
    enabled: true
    size: 30Gi  #SSD Required
    storageClass:
  replicaCount: 3
  resources:
    limits:
      cpu: 4
      memory: 8Gi
    requests:
      cpu: 4
      memory: 8Gi
* externalS3:
  enabled: true

* externalKafka: //I use AWS MSK service. I used two broker(`kafka.m5.xlarge,` vCPU: 4, Memory: 16Gi)
  enabled: true
```

The following latency was measured for searching 10M 256 dim vectors.(top 100)

![image](https://user-images.githubusercontent.com/65549245/217556720-dd18563c-a043-4882-a30d-ee2e9e77753a.png)

In my opinion, `Search in queue latency` accounts for a large portion of total latency. I want to reduce it, but I don't know which part to address.
Does the queryNode's cpu or memory affect `search in queue latency`? Or is it spec of kafka broker? Or are there other factors?

**Comment by yhmo** [comment, marked as answer]
Search performance is affected by many factors. When we say performance, we often talk about qps and latency.
Major factors affect search performance:
- data properties: rows, dimension
- vector index, make sure data is fully indexed
- how many query nodes, large dataset requires more nodes
- hardware ability of query node, cpu cores
- client-side behavior,  search request parameters

There is a request queue for each query node.
The "Search request latency" is the total time cost of a request from begin to end.
The "Search in queue latency" is the time cost of a request waiting in the queue. 
The "Search segment latency" is the time cost in ANN computation.

"Search request latency" = "Search in queue latency" + "Search segment latency"

**Comment by yhmo** [comment]
Milvus loads segments into query nodes, and try to keep load balance between nodes. If there are only a few segments, the balance is not good.
There is a metric "Estimated CPU Usage", if the usage is low, try reduce query nodes.
4 query nodes with cpu:4, memory: 16Gi ----> 2 query nodes with cpu:8, memory: 32Gi

**Reply by redlion0929** [reply]
Thank you for answering.
Then, if i scale up kafka or change s3 to minio, it brings few impact for search performance(latency). Is it right?

**Reply by yhmo** [reply]
Kafka/Pulsar acts as a WAL component. When Proxy receives insert/delete requests, it passes the data into Kafka/Pulsar to avoid data loss. Then datanode/querynode subscribe data from Kafka/Pulsar. Datanode persists the data into storage(S3/MinIO); Querynode keeps the data in memory for searching.
Kafka/Pulsar has almost no impact on search performance, and only affects data visibility/consistency in streaming data scenarios. In this kind of scenario, there is a parameter "consistency_level", set this parameter to "Strong" if you need strong consistency, the querynode will wait until all the data is consumed from Kafka/Pulsar so that the search action can see the entire dataset.

Milvus stores data files in MinIO/S3, it requires all data of the collection to be loaded into memory before searching. The bandwidth of S3/MinIO only affect the load performance. Once all the data is loaded, the search action will use the in-memory data to compute. So, S3/MinIO has no impact to search performance.

**Comment by redlion0929** [comment]
Thank you for answering.
Actually, I changed the consistency of collection and search from `Strong` to `Bounded` today.
Now, the latency reduces 200ms to 5ms. 
Because the default value of creating Collection of milvus sdk go is `Strong`, I couldn't consider it.
I was surprised that the consistency setting has a big impact than I think of the search latency.

**Reply by yhmo** [reply]
Yes, because "Strong" consistency need to wait a "time tick" to ensure data is consumed from pulsar/kafka.
In milvus core, the "time tick" is exactly equal to 200ms.
